docs/proposals/configmap-rollout.md

```diff
@@ -0,0 +1,181 @@
+deployment "bar" updated
+```
+
+A new controller needs to watch for Deployments and ConfigMaps. Whenever it
+spots a Deployment with a `triggered-by` annotation and a mounted ConfigMap,
+it will create a copy of the ConfigMap and mount the copy in the Deployment.
+It should generally track Deployments with `triggered-by` annotations, 
+create new ConfigMap copies on updates of the original ConfigMap, and mount
+them back in the Deployment. Only Deployments should work for this controller
+for now but in the future the controller can be extended to facilitate config 
+management for DaemonSets, PetSets, and any other resource that may need it.
+
+Once the copied ConfigMap is mounted in the Deployment, the deployment
+controller needs to create the new ReplicaSet and set the owner reference
+for the copy to point to the ReplicaSet. In that way we ensure that copies
+of the original ConfigMap will be pruned by the garbage collector once their
+owner ReplicaSets are deleted. Certain copies of a ConfigMap will end up 
+being shared by more than one ReplicaSets.
+
+Additionally, the deployment controller should not create a new ReplicaSet
+for a Deployment "bar" that holds a `triggered-by` annotation to a ConfigMap
+"foo". Instead, it should wait until the new controller updates the Deployment
+template with the first copy of "foo". This ensures that the original ConfigMap
+copy will not land in a ReplicaSet, hence will not be subject to garbage
+collection.
+
+
+### Alternative solutions
+
+* Embed ConfigMap as a template in the Deployment.
+
+While it will work smoothly on ConfigMap changes for Deployments, this
+solution will need to be spread across the API for all resources that face
+the same problem. Sooner or later, higher-level resources will get bloated
+trying to manage stuff like ConfigMaps, Secrets, et al.
+
+* Others?
+
+## Changes
+
+### API
+
+No API changes will be needed for Deployments or ConfigMaps.
+
+### Controller
+
+A new controller will be needed for watching Deployments and Config Maps. Changes will also be needed in the deployment controller. [See above](#proposed-solution).
+
+### CLI
+
+We need a way to specify relationship between a resource that is updated (ConfigMap in our case) and a resource that will be triggered on those updates (Deployment). A generic `set trigger` will be added in `kubectl`. The relationship will be established with an annotation on the triggered resource (Deployment).
+```sh
+$ kubectl set trigger deployment/bar configmap/foo
+deployment "bar" updated
+$ kubectl get deployment/bar -o yaml | grep triggered-by
+deployment.kubernetes.io/triggered-by: configmap/foo
+```
+
+We also need a way to set volumes in pod templates without having to edit files by hand. A `kubectl set volume` command will be added to accommodate that need.
+```sh
+$ kubectl set volume deployment/bar -t configmap --configmapname foo
+deployment "bar" updated
+```
+
+## Issues
```

Comment:
"pause, update config, update image, resume" is not better than "create config map, update deployment, delete old config map". If I'm going to write that level of automation, I don't need the system proposed here.

Regarding incompatibilities: I think we're talking about two kinds of incompatibilities. You're probably thinking of pods that have a different external interface. I am considering the interface between pod and config map. In my view, the config map is an integral part of a version of the software, and rolled out together with it. This doesn't need to be a backwards- or forwards-compatible interface.

Let me explain a bit of background: We had a long and painful history with command-line flags at Google. During pushes, a new flag is introduced but not set correctly, or an old flag is removed but still set. We fixed this with the concept of "config MPM", where the flags would be bundled together with the software and rolled out as a unit. As config maps are Kubernetes' answer to flags (i.e., information that's not built into the image but injected by the deployment system), I think we should follow similar best practices.